cmd/zhack/zhack.c

```diff
@@ -469,6 +481,326 @@ zhack_do_feature(int argc, char **argv)
+
+	// Scan the entire DDT
+	zfs_scrub_ddt_class_max = DDT_CLASS_UNIQUE;
+
+	// Disable reference tracking debugging
+	reference_tracking_enable = B_FALSE;
+
+	// Disable prefetch during scan
+	zfs_no_scrub_prefetch = B_TRUE;
+
+	while ((c = getopt(argc, argv, "D:EG:H:PRTi:S:nrv")) != -1) {
+		switch (c) {
+		case 'D':
+			// How much of the DDT are we scanning?
+		{
+			char *endptr = NULL;
+			unsigned long class =
+			    strtoul(optarg, &endptr, 0);
+			if ((errno == 0) && (*endptr == '\0') &&
+			    (class < DDT_CLASSES)) {
+				zfs_scrub_ddt_class_max = class;
+			} else {
+				fatal(NULL, FTAG, "DDT class must be between "
+				    "0 and %d, inclusive", DDT_CLASSES-1);
+			}
+		}
+		break;
+		case 'E':
+			// Forcibly reset DDT class max after import
+			do_ddt_reset++;
+			break;
+		case 'G':
+		{
+			char *endptr = NULL;
+			uint64_t gap = strtoul(optarg, &endptr, 0);
+			if ((errno == 0) && (*endptr == '\0')) {
+				zfs_scan_max_ext_gap = gap;
+			} else {
+				fatal(NULL, FTAG, "Bad range tree gap (-G)");
+			}
+		}
+		break;
+		case 'H':
+		{
+			char *endptr = NULL;
+			uint64_t fact = strtoul(optarg, &endptr, 0);
+			if ((errno == 0) && (*endptr == '\0') &&
+			    (fact >= 1) && (fact <= 1000)) {
+				zfs_scan_mem_lim_fact = fact;
+			} else {
+				fatal(NULL, FTAG, "Bad hard factor (-H)");
+			}
+		}
+		break;
+		case 'M':
+		{
+			char *endptr = NULL;
+			uint64_t mem = strtoul(optarg, &endptr, 0);
+			if ((errno == 0) && (*endptr == '\0') &&
+			    (mem >= (1 << 15)) &&
+			    (mem <= sysconf(_SC_PHYS_PAGES))) {
+				physmem_override = mem;
+			} else {
+				fatal(NULL, FTAG,
+				    "Bad physical memory override (-M)");
+			}
+		}
+		break;
+		case 'P':
+			// Turn prefetching back on
+			zfs_no_scrub_prefetch = B_FALSE;
+			break;
+		case 'R':
+			// Restart a scrub
+			do_restart++;
+			break;
+		case 'S':
+		{
+			char *endptr = NULL;
+			uint64_t fact = strtoul(optarg, &endptr, 0);
+			if ((errno == 0) && (*endptr == '\0') &&
+			    (fact >= 1) &&
+			    (fact <= 1000)) {
+				zfs_scan_mem_lim_soft_fact = fact;
+			} else {
+				fatal(NULL, FTAG, "Bad soft factor (-S)");
+			}
+		}
+		break;
+		case 'T':
+			// Turn reference tracking back on
+			reference_tracking_enable = B_TRUE;
+			break;
+		case 'i':
+		{
+			char *endptr = NULL;
+			uint64_t intval = strtoul(optarg, &endptr, 0);
+			if ((errno == 0) && (*endptr == '\0')) {
+				zfs_scan_checkpoint_intval = intval;
+			} else {
+				fatal(NULL, FTAG, "Bad scan interval (-i)");
+			}
+		}
+		break;
+		case 'n':
+			// Don't launch a scrub, just resume one
+			no_spawn++;
+			break;
+		case 'r':
+			// Resilver instead of scrub
+			do_resilver++;
+			break;
+		case 'v':
+			// Be chatty
+			verbose++;
+			break;
+		case '?':
+			fatal(NULL, FTAG, "invalid option '%c'", optopt);
+		}
+	}
+
+	if (optind == argc) {
+		fatal(NULL, FTAG, "Need pool name");
+	}
+	if (optind + 1 < argc) {
+		(void) fprintf(stderr,
+		    "WARNING: Discarding excess arguments\n");
```

Comment:
For consistency with the other utilities this should be fatal and print the usage on failure.